department-of-veterans-affairs/caseflow
# Add export appeal endpoint for System Admins

### Description
Extends #15976 to allow exporting appeals and associated records via the browser. This avoids having logging into prod, running export commands, and scp-ing the file locally.
The endpoint only works for System Admins (`user.admin? => true`).

Bonus 1: shows task tree and intake rendering in the browser
Bonus 2: shows timeline and network graph visualizations

### Acceptance Criteria
- [ ] JSON file contents can be downloaded from the browser

Added documentation: https://github.com/department-of-veterans-affairs/caseflow/wiki/Exporting-and-Importing-Appeals

### Testing Plan
1. In dev Rails console, list System Admin users: `Functions.list_all`
2. In Caseflow, log in as a System Admin (e.g., `BVASCASPER1`)
3. Go to their queue: http://localhost:3000/queue and click on a case to open Case Details. Try with AMA and legacy cases.
4. Change the url from `queue` to `export`, like this: `http://localhost:3000/export/appeals/31927be1-8447-44c4-ad66-8b73ae8bf401`
5. Click on the blue 'Help' text
   - Append `?show_pii`to the URL like this: `http://localhost:3000/export/appeals/31927be1-8447-44c4-ad66-8b73ae8bf401?show_pii` or click on the link
   - Customize task tree columns like this: `http://localhost:3000/export/appeals/31927be1-8447-44c4-ad66-8b73ae8bf401?fields=id,status,updated_at,CLO_DATE` or click on the link for an example
   - Click on the `json` link to show the JSON file contents, which can then be examined for PII and then saved locally for importing.

Repeat the above logged in as a non-admin. Should see an error:
```
{"errors":[{"status":403,"title":"Additional access needed","detail":"Additional access needed"}]}
```

You can also add `binding.pry` in `export_spec.rb` and examine the page.

### Screenshots
![image](https://user-images.githubusercontent.com/55255674/112224053-c3ea6200-8bf8-11eb-903b-2ee5405a12a7.png)
![image](https://user-images.githubusercontent.com/55255674/112227334-7d4b3680-8bfd-11eb-8ed5-d9f1b1a70df3.png)

![image](https://user-images.githubusercontent.com/55255674/116741964-5cc08a00-a9bc-11eb-9971-898dad72ed06.png)
